python/GafferUI/CodeWidget.py

```diff
@@ -0,0 +1,577 @@
+
+		self.__commentPrefix = commentPrefix
+
+	def getCommentPrefix( self ) :
+
+		return self.__commentPrefix
+
+	def __keyPress( self, widget, event ) :
+
+		if event.key == "Tab" :
+			if self._qtWidget().textCursor().hasSelection() :
+				if event.modifiers == event.Modifiers.None_ :
+					return self.__indentPress( 1 )
+				elif event.modifiers == event.Modifiers.Shift :
+					return self.__indentPress( -1 )
+			elif event.modifiers == event.Modifiers.None_ :
+				return self.__completionPress()
+		elif event.key == "Backtab" :
+			return self.__indentPress( -1 )
+		elif event.modifiers == event.Modifiers.Control and event.key == "BracketLeft" :
+			return self.__indentPress( -1 )
+		elif event.modifiers == event.Modifiers.Control and event.key == "BracketRight" :
+			return self.__indentPress( 1 )
+		elif event.modifiers == event.Modifiers.Control and event.key == "Slash" :
+			return self.__commentPress()
+		elif event.key == "Return" and event.modifiers == event.Modifiers.None_ :
+			return self.__returnPress()
+
+		return False
+
+	def __completionsAndLine( self ) :
+
+		if self.__completer is None :
+			return [], ""
+
+		text = self.getText()
+		cursor = self.getCursorPosition()
+		lineStart = max( text.rfind( "\n", 0, cursor ), 0 )
+		line = text[lineStart:cursor]
+
+		return self.__completer.completions( line ), line
+
+	def __completionPress( self ) :
+
+		completions, line = self.__completionsAndLine()
+		if not completions :
+			return False
+
+		commonPrefix = os.path.commonprefix( [ c.text for c in completions ] )
+		if len( commonPrefix ) > len( line ) :
+			self.insertText( commonPrefix[len(line):] )
+
+		return bool( completions )
+
+	def __selectedBlocks( self ) :
+
+		cursor = self._qtWidget().textCursor()
+		startBlock = self._qtWidget().document().findBlock( cursor.selectionStart() )
+		endBlock = self._qtWidget().document().findBlock( cursor.selectionEnd() )
+
+		result = []
+		while True :
+			result.append( startBlock )
+			if startBlock == endBlock :
+				break
+			else :
+				startBlock = startBlock.next()
+
+		if len( result ) > 1 and result[-1].position() == cursor.selectionEnd() :
+			# Omit last block if cursor is at the start of the line.
+			del result[-1]
+
+		return result
+
+	def __indentPress( self, indent ) :
+
+		cursor = self._qtWidget().textCursor()
+		try :
+
+			cursor.beginEditBlock()
+
+			for block in self.__selectedBlocks() :
+
+				cursor = QtGui.QTextCursor( block )
+				if indent > 0 :
+					cursor.insertText( "\t" * indent )
+				else :
+					for i in range( 0, len( os.path.commonprefix( [ block.text(), "\t" * -indent ] ) ) ) :
+						cursor.deleteChar()
+
+		finally :
+
+			cursor.endEditBlock()
+
+		return True
+
+	def __commentPress( self ) :
+
+		if not self.__commentPrefix :
+			return False
+
+		# Capture groups :
+		#
+		# 1) optional indent
+		# 2) optional comment prefix with optional trailing space
+		regex = re.compile( r"^(\t*)({} ?)?".format( self.__commentPrefix ) )
```

Comment:
I honestly find that less readable - now the opening and closing of the last capture group are in two separate strings, separated by a bunch of noise. In https://github.com/GafferHQ/gaffer/pull/4314/commits/f6bbf70e220ae9a1b0fd975471042c5428d99975 I've used a named format arg which hopefully makes it clearer for you - you didn't seem to mind me using that to build up the more complex regexes elsewhere?

I'm trying to use `str.format()` consistently for pretty much everything now, although I still waver a bit when its just a simple `"prefix{}".format( suffix )`. Where it will really be a win is when we can finally drop support for Python 2 and convert to [f-strings](https://realpython.com/python-f-strings/#simple-syntax).